Src/PerformanceCollector/Shared/QuickPulseTelemetryModule.cs

```diff
@@ -0,0 +1,468 @@
+                        pc.InstanceName,
+                        false,
+                        true);
+
+                    QuickPulseEventSource.Log.CounterRegisteredEvent(counter.Item2);
+                }
+                catch (InvalidOperationException e)
+                {
+                    QuickPulseEventSource.Log.CounterRegistrationFailedEvent(e.Message, counter.Item2);
+                }
+            }
+        }
+
+        private void ValidateConfiguration(TelemetryConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            if (configuration.TelemetryInitializers == null)
+            {
+                throw new ArgumentNullException(nameof(configuration.TelemetryInitializers));
+            }
+        }
+
+        private void InitializeTimers()
+        {
+            this.collectionTimer = new Timer(this.CollectionTimerCallback);
+
+            this.stateTimer = new Timer(this.StateTimerCallback);
+            this.stateTimer.ScheduleNextTick(this.initialDelay);
+        }
+
+        private IQuickPulseTelemetryProcessor FetchTelemetryProcessor(TelemetryConfiguration configuration)
+        {
+            return configuration.TelemetryProcessors.OfType<IQuickPulseTelemetryProcessor>().SingleOrDefault();
+        }
+
+        private void InitializeServiceClient(TelemetryConfiguration configuration)
+        {
+            if (this.serviceClient != null)
+            {
+                // service client has been passed through a constructor, we don't need to do anything
+                return;
+            }
+
+            Uri serviceEndpointUri;
+            if (string.IsNullOrWhiteSpace(this.QuickPulseServiceEndpoint))
+            {
+                // endpoint is not specified in configuration, use the default one
+                serviceEndpointUri = this.serviceUriDefault;
+            }
+            else
+            {
+                // endpoint appears to have been specified in configuration, try using it
+                try
+                {
+                    serviceEndpointUri = new Uri(this.QuickPulseServiceEndpoint);
+                }
+                catch (Exception e)
+                {
+                    throw new ArgumentException(
+                        string.Format(
+                            CultureInfo.InvariantCulture,
+                            "Error initializing QuickPulse module. QPS endpoint is not a correct URI: '{0}'",
+                            this.QuickPulseServiceEndpoint),
+                        e);
+                }
+            }
+
+            // create the default production implementation of the service client with the best service endpoint we could get
+            this.serviceClient = new QuickPulseServiceClient(serviceEndpointUri, GetInstanceName(configuration), SdkVersionUtils.GetAssemblyVersion());
```

Comment:
It may be too early to read instance name as customer may set it's own initializer for role instance name population